src/wagtail_live/adapters/telegram/receiver.py

```diff
@@ -0,0 +1,273 @@
+
+        if command_text == "/get_chat_id":
+            chat_id = self.get_channel_id_from_message(message=message)
+            response = requests.get(
+                get_base_telegram_url() + "sendMessage",
+                params={
+                    "chat_id": chat_id,
+                    "text": chat_id,
+                },
+            )
+
+            payload = response.json()
+            if not payload["ok"]:
+                logger.error(payload["description"])
+
+    def get_channel_id_from_message(self, message):
+        """See base class."""
+
+        # Since live posts aren't cleaned when they are added via a receiver,
+        # we make sure at this level that we return the correct types.
+        return str(message["chat"]["id"])
+
+    def get_message_id_from_message(self, message):
+        """Messages containing multiple images have a media_group_id attribute.
+        See base class.
+        """
+
+        msg_id = message.get("media_group_id") or message.get("message_id")
+        return str(msg_id)
+
+    def get_message_text(self, message):
+        """See base class."""
+
+        # Telegram parses the text of a message before sending it.
+        # The result can be found in the message's "entities".
+        return {
+            "text": message.get("text") or message.get("caption") or "",
+            "entities": message.get("entities", []),
+        }
+
+    def process_text(self, live_post, message_text):
+        """Use the message entities to convert links.
+        A raw link isn't converted by Telegram.
+        A link with a description is sent as a `text_link` entity.
+        See base class.
+        """
+
+        text = message_text["text"]
+        len_text = len(text)
+
+        cumulated_offset = 0
+        entities = message_text["entities"]
+        for entity in entities:
+            url = ""
+            offset = int(entity["offset"]) + cumulated_offset
+            length = int(entity["length"])
+            end = offset + length
+
+            if entity["type"] == "url":
+                url = description = text[offset:end]
+                if is_embed(url):
+                    # Check if this can match an embed block, if so no conversion happens.
+                    if end == len_text or (
+                        len_text > end + 1 and text[end : end + 1] == "\n"  # noqa
+                    ):
+                        if offset == 0 or text[offset - 1 : offset] == "\n":  # noqa
+                            # This is an embed block
+                            continue
+
+            if entity["type"] == "text_link":
+                url = entity["url"]
+                description = text[offset:end]
+
+            if url:
+                link = f"<a href='{url}'>{description}</a>"
+                text = text[:offset] + link + text[end:]
+                len_added = len(link) - length
+                cumulated_offset += len_added
+                len_text += len_added
```

Comment:
I think this manual algorithm is a bit too complicated which makes it hard to understand what happens. After some thinking I concluded that no extra variables are needed if the entities are process backwards, because the text will then only be edited at locations which are no longer matches by indices in the entity objects.

I propose the following algorithm:

```python
# Process the entities in reversed order to be able to edit the text in place.
for entity in reversed(entities):
    url = ""
    start = entity['offset']
    end = start + entity['length']

    if entity["type"] == "url":
        url = description = text[start:end]

        if is_embed(url):
            # Check if this can match an embed block, if so no conversion happens.
            # It matches an embed block if it has a line in the text for itself.
            if end == len_text or text[end] == "\n":
                if start == 0 or text[start - 1] == "\n":
                    # This is an embed block, skip to the next entity
                    continue

    if entity["type"] == "text_link":
        url = entity["url"]
        description = text[start:end]

    if url:
        link = f'<a href="{url}">{description}</a>'
        text = text[:start] + link + text[end:]
```

Reply:
Very nice algorithm. Thanks!